expression/builtin_encryption.go

```diff
@@ -895,10 +897,165 @@ func (b *builtinUncompressedLengthSig) evalInt(row chunk.Row) (int64, bool, erro
+	}
+	if s.User == nil {
+		return true, nil
+	}
+	if n := s.User.Username; n != "" {
+		if psw == n || psw == reverse(n) {
+			return false, nil
+		}
+	}
+	if n := s.User.AuthUsername; n != "" {
+		if psw == n || psw == reverse(n) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func validateByMixedDigitSpecial(s *variable.SessionVars, pwd string) (bool, error) {
+	// get system variables
+	v, err := variable.GetGlobalSystemVar(s, variable.ValidatePasswordMixedCaseCount)
+	if err != nil {
+		return false, err
+	}
+	minMixed, err := strconv.ParseInt(v, 10, 64)
+	if err != nil {
+		return false, err
+	}
+	v, err = variable.GetGlobalSystemVar(s, variable.ValidatePasswordNumberCount)
+	if err != nil {
+		return false, err
+	}
+	minDigit, err := strconv.ParseInt(v, 10, 64)
+	if err != nil {
+		return false, err
+	}
+	v, err = variable.GetGlobalSystemVar(s, variable.ValidatePasswordSpecialCharCount)
+	if err != nil {
+		return false, err
+	}
+	minSpecial, err := strconv.ParseInt(v, 10, 64)
+	if err != nil {
+		return false, err
+	}
+
+	numLower := int64(0)
+	numUpper := int64(0)
+	numDigit := int64(0)
+	numSpecial := int64(0)
+	for _, c := range pwd {
+		if unicode.IsLower(c) {
+			numLower++
+		} else if unicode.IsUpper(c) {
+			numUpper++
+		} else if unicode.IsDigit(c) {
+			numDigit++
+		} else {
+			numSpecial++
+		}
+	}
+
+	if numLower < minMixed || numUpper < minMixed || numDigit < minDigit || numSpecial < minSpecial {
+		return false, nil
+	}
+	return true, nil
+}
+
+// TODO: Support validating password by dictionary file
+func validateByDictionary(s *variable.SessionVars, pwd string) (bool, error) {
+	return true, nil
+}
+
 type validatePasswordStrengthFunctionClass struct {
 	baseFunctionClass
 }
 
 func (c *validatePasswordStrengthFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "VALIDATE_PASSWORD_STRENGTH")
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, types.ETString)
+	bf.tp.Flag = mysql.MaxIntWidth
+	sig := &builtinValidatePasswordStrengthSig{bf}
+	return sig, nil
+}
+
+type builtinValidatePasswordStrengthSig struct {
+	baseBuiltinFunc
+}
+
+func (c *builtinValidatePasswordStrengthSig) Clone() builtinFunc {
+	newSig := &builtinValidatePasswordStrengthSig{}
+	newSig.cloneFrom(&c.baseBuiltinFunc)
+	return newSig
+}
+
+// evalInt evals VALIDATE_PASSWORD_STRENGTH(str).
+// See https://dev.mysql.com/doc/refman/8.0/en/encryption-functions.html#function_validate-password-strength
+func (c *builtinValidatePasswordStrengthSig) evalInt(row chunk.Row) (int64, bool, error) {
+	sv := c.ctx.GetSessionVars()
+	pwd, isNull, err := c.args[0].EvalString(c.ctx, row)
+	score := int64(0)
+
+	if isNull || err != nil {
+		return score, true, err
+	}
+
+	// In MySQL, the max length of a password is 100
+	l := int64(0)
+	for index := range pwd {
```

Comment:
1. Please give a name for the magic number `100` and use that variable name. For example:
```go
var maxPwdLength int64 = 100
...
func (c * builtinValidatePasswordStrengthSig) evalInt(...) (...) {
	...
	if index > maxPwdLength { ... }
	...
}
```

2. Do not a `for` loop to check whether the length of `pwd` is larger than 100. You can:
```go
l := len(pwd)
if l > 100 {
	l = 100
}
pwd = pwd[:l]
```